drivers/mspi/mspi_stm32_qspi.c

```diff
@@ -0,0 +1,1382 @@
+		LOG_ERR("failed to change device cfg");
+	}
+
+	/* Release PM resources */
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	(void)pm_device_runtime_put(controller);
+
+e_return:
+	if (locked) {
+		k_mutex_unlock(&data->lock);
+	}
+
+	return ret;
+}
+
+/**
+ * API implementation of mspi_xip_config : XIP configuration
+ *
+ * @param controller Pointer to the device structure for the driver instance.
+ * @param dev_id Pointer to the device ID structure from a device.
+ * @param xip_cfg The controller XIP configuration for MSPI.
+ *
+ * @retval 0 if successful.
+ * @retval -ESTALE device ID don't match, need to call mspi_dev_config first.
+ */
+static int mspi_stm32_qspi_xip_config(const struct device *controller,
+				      const struct mspi_dev_id *dev_id,
+				      const struct mspi_xip_cfg *xip_cfg)
+{
+	struct mspi_stm32_data *dev_data = controller->data;
+	int ret = 0;
+
+	if (dev_id != dev_data->dev_id) {
+		LOG_ERR("xip_config: dev_id don't match");
+		return -ESTALE;
+	}
+
+	ret = pm_device_runtime_get(controller);
+	if (ret != 0) {
+		LOG_ERR("%u, pm_device_runtime_get() failed: %d", __LINE__, ret);
+		return ret;
+	}
+
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	if (!xip_cfg->enable) {
+		/* This is for aborting memory mapped mode */
+		ret = mspi_stm32_qspi_memmap_off(controller);
+	} else {
+		ret = mspi_stm32_qspi_memmap_on(controller);
+	}
+
+	if (ret == 0) {
+		dev_data->xip_cfg = *xip_cfg;
+		LOG_INF("QSPI XIP configured %d", xip_cfg->enable);
+	}
+
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	if (pm_device_runtime_put(controller)) {
+		LOG_ERR("%u, pm_device_runtime_put() failed", __LINE__);
+	}
+
+	return ret;
+}
+
+/**
+ * API implementation of mspi_get_channel_status.
+ *
+ * @param controller Pointer to the device structure for the driver instance.
+ * @param ch Not used.
+ *
+ * @retval 0 if successful.
+ * @retval -EBUSY MSPI bus is busy
+ */
+static int mspi_stm32_qspi_get_channel_status(const struct device *controller, uint8_t ch)
+{
+	struct mspi_stm32_data *data = controller->data;
+	QSPI_HandleTypeDef *hmspi = &data->hmspi.qspi;
+	int ret = 0;
+
+	ARG_UNUSED(ch);
+
+	if (mspi_is_inp(controller) || (hmspi->Instance->SR & QUADSPI_SR_BUSY) != 0) {
+		ret = -EBUSY;
+	}
+
+	data->dev_id = NULL;
+
+	return ret;
+}
+
+static int mspi_stm32_qspi_pio_transceive(const struct device *controller,
+					  const struct mspi_xfer *xfer)
+{
+	struct mspi_stm32_data *dev_data = controller->data;
+	struct mspi_stm32_context *ctx = &dev_data->ctx;
+	const struct mspi_xfer_packet *packet;
+	uint32_t packet_idx;
+	int ret = 0;
+
+	if (xfer->num_packet == 0 || xfer->packets == NULL ||
+	    xfer->timeout > CONFIG_MSPI_COMPLETION_TIMEOUT_TOLERANCE) {
+		LOG_ERR("Transfer: wrong parameters");
+		return -EFAULT;
+	}
+
+	/* DummyCycle to give to the mspi_stm32_read_access/mspi_stm32_write_access */
+	ret = mspi_context_lock(ctx, dev_data->dev_id, xfer, true);
```

Comment:
`mspi_context_lock()` is only called from here. I would suggest you remove the function and implement instrcution here. It would be easier to see that here we take a semaphore and on function exit we give it back.

Alternatively, move the `k_sem_give()` instruction into a local `mspi_context_unlock()` helper function .